// @apps-in-toss/web-framework@1.6.1/dist/config/index.js
// ../../../.yarn/cache/es-toolkit-npm-1.32.0-019fc7c155-56ba965570.zip/node_modules/es-toolkit/dist/predicate/isPlainObject.mjs
function isPlainObject(value) {
  if (!value || typeof value !== "object") {
    return false;
  }
  const proto = Object.getPrototypeOf(value);
  const hasObjectPrototype = proto === null || proto === Object.prototype || Object.getPrototypeOf(proto) === null;
  if (!hasObjectPrototype) {
    return false;
  }
  return Object.prototype.toString.call(value) === "[object Object]";
}

// ../../../.yarn/cache/es-toolkit-npm-1.32.0-019fc7c155-56ba965570.zip/node_modules/es-toolkit/dist/object/merge.mjs
function merge(target, source) {
  const sourceKeys = Object.keys(source);
  for (let i = 0; i < sourceKeys.length; i++) {
    const key = sourceKeys[i];
    const sourceValue = source[key];
    const targetValue = target[key];
    if (Array.isArray(sourceValue)) {
      if (Array.isArray(targetValue)) {
        target[key] = merge(targetValue, sourceValue);
      } else {
        target[key] = merge([], sourceValue);
      }
    } else if (isPlainObject(sourceValue)) {
      if (isPlainObject(targetValue)) {
        target[key] = merge(targetValue, sourceValue);
      } else {
        target[key] = merge({}, sourceValue);
      }
    } else if (targetValue === void 0 || sourceValue !== void 0) {
      target[key] = sourceValue;
    }
  }
  return target;
}

// src/config/defineConfig.ts
var defineConfig = (config) => {
  const defaultConfig = {
    scheme: "intoss",
    outdir: "dist",
    web: {
      host: "localhost"
    },
    webViewProps: {
      type: "partner",
      mediaPlaybackRequiresUserAction: true,
      allowsInlineMediaPlayback: false,
      bounces: true,
      allowsBackForwardNavigationGestures: true,
      pullToRefreshEnabled: true,
      overScrollMode: "always"
    },
    navigationBar: {
      withBackButton: true,
      withHomeButton: false
    }
  };
  return merge(defaultConfig, config);
};
export {
  defineConfig
};
